server/etcdserver/api/v3discovery/discovery_test.go

```diff
@@ -0,0 +1,651 @@
+}
+
+func TestGetClusterMembers(t *testing.T) {
+	lg, err := zap.NewProduction()
+	if err != nil {
+		t.Errorf("Failed to create a logger, error: %v", err)
+	}
+
+	d := &discovery{
+		lg: lg,
+		c: &clientv3.Client{
+			KV: &fakeKVForClusterMembers{
+				fakeBaseKV: &fakeBaseKV{},
+				members:    actualMemberInfo,
+			},
+		},
+		cfg:          &DiscoveryConfig{},
+		clusterToken: clusterFakeToken,
+	}
+
+	clsInfo, _, err := d.getClusterMembers()
+	if err != nil {
+		t.Errorf("Failed to get cluster members, error: %v", err)
+	}
+
+	if clsInfo.Len() != len(expectedMemberInfo) {
+		t.Errorf("unexpected member count, expected: %d, got: %d", len(expectedMemberInfo), clsInfo.Len())
+	}
+
+	for i, m := range clsInfo.members {
+		if m != expectedMemberInfo[i] {
+			t.Errorf("unexpected member[%d], expected: %v, got: %v", i, expectedMemberInfo[i], m)
+		}
+	}
+}
+
+// fakeKVForCheckCluster is used to test checkCluster.
+type fakeKVForCheckCluster struct {
+	*fakeBaseKV
+	t                 *testing.T
+	token             string
+	clusterSizeStr    string
+	members           []memberInfo
+	getSizeRetries    int
+	getMembersRetries int
+}
+
+// We only need to overwrite method `Get`.
+func (fkv *fakeKVForCheckCluster) Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	clusterSizeKey := fmt.Sprintf("/_etcd/registry/%s/_config/size", fkv.token)
+	clusterMembersKey := fmt.Sprintf("/_etcd/registry/%s/members", fkv.token)
+
+	if key == clusterSizeKey {
+		if fkv.getSizeRetries > 0 {
+			fkv.getSizeRetries--
+			// discovery client should retry on error.
+			return nil, errors.New("get cluster size failed")
+		}
+		return &clientv3.GetResponse{
+			Kvs: []*mvccpb.KeyValue{
+				{
+					Value: []byte(fkv.clusterSizeStr),
+				},
+			},
+		}, nil
+
+	} else if key == clusterMembersKey {
+		if fkv.getMembersRetries > 0 {
+			fkv.getMembersRetries--
+			// discovery client should retry on error.
+			return nil, errors.New("get cluster members failed")
+		}
+		kvs := memberInfoToKeyValues(fkv.members)
+
+		return &clientv3.GetResponse{
+			Header: &etcdserverpb.ResponseHeader{
+				Revision: 10,
+			},
+			Kvs: kvs,
+		}, nil
+	} else {
+		fkv.t.Errorf("unexpected key: %s", key)
+		return nil, fmt.Errorf("unexpected key: %s", key)
+	}
+}
+
+func TestCheckCluster(t *testing.T) {
+	cases := []struct {
+		memberId          types.ID
+		getSizeRetries    int
+		getMembersRetries int
+		expectedError     error
+	}{
+		{
+			memberId:          101,
+			getSizeRetries:    0,
+			getMembersRetries: 0,
+			expectedError:     nil,
+		},
+		{
+			memberId:          102,
+			getSizeRetries:    2,
+			getMembersRetries: 0,
+			expectedError:     nil,
+		},
+		{
+			memberId:          103,
+			getSizeRetries:    0,
+			getMembersRetries: 2,
+			expectedError:     nil,
+		},
+		{
+			memberId:          104,
+			getSizeRetries:    0,
+			getMembersRetries: 0,
+			expectedError:     ErrFullCluster,
+		},
+	}
+
+	for i, tc := range cases {
```

Comment:
Please use subtest, https://ieftimov.com/post/testing-in-go-subtests/

Reply:
I agree using subtest is better. Previously I thought it isn't a big deal, so I did not use it. 

Will resolve it, thanks.

Reply:
Resolved & thanks.